scipy/special/specfun/specfun.f

```diff
@@ -10007,112 +10007,120 @@ SUBROUTINE JYZO(N,NT,RJ0,RJ1,RY0,RY1)
         DIMENSION RJ0(NT),RJ1(NT),RY0(NT),RY1(NT)
         PI=3.141592653589793D0
 C       -- Newton method for j_{N,L}
-C       1) initial guess for j_{N,1}
-        IF (N.LE.20) THEN
-           X=2.82141+1.15859*N
+C       initial guess for j_{N,1}
+        IF (N.EQ.0) THEN
+           X=2.4
         ELSE
 C          Abr & Stg (9.5.14)
            X=N+1.85576*N**0.33333+1.03315/N**0.33333
         ENDIF
+C       iterate
         L=0
-C       2) iterate
-        XGUESS=X
 10      X0=X
         CALL JYNDD(N,X,BJN,DJN,FJN,BYN,DYN,FYN)
         X=X-BJN/DJN
-        IF (X-X0.LT.-1) X=X0-1
-        IF (X-X0.GT.1) X=X0+1
         IF (DABS(X-X0).GT.1.0D-11) GO TO 10
-C       3) initial guess for j_{N,L+1}
-        IF (L.GE.1)THEN
-           IF (X.LE.RJ0(L)+0.5) THEN
-              X=XGUESS+PI
-              XGUESS=X
-              GO TO 10
-           ENDIF
-        END IF
         L=L+1
         RJ0(L)=X
-C       XXX: should have a better initial guess for large N ~> 100 here
-        X=X+PI+MAX((0.0972d0+0.0679*N-0.000354*N**2)/L, 0d0)
+C       initial guess for j_{N,L}
+        IF (L.EQ.1) THEN
+           IF (N.EQ.0) THEN
+              X=5.52
+           ELSE
+C             Expansion from https://dlmf.nist.gov/10.21#E32 and
+C             coefficients from Olver 1951
+              X=N+3.24460*N**0.33333+3.15824/N**0.33333
+           ENDIF
+        ELSE
+C          growth of roots is approximately linear (https://dlmf.nist.gov/10.21#E20)
+           X=RJ0(L) + (RJ0(L) - RJ0(L-1))
+        ENDIF
+        IF (L.LE.N+10) THEN
+           CALL JYNDD(N,X,BJN,DJN,FJN,BYN,DYN,FYN)
+           H=DATAN(DABS(DJN)/DSQRT(DABS(FJN * BJN)))
+           B=-DJN / (BJN * DTAN(H))
+           X=X - (H - PI/2) / B
+        ENDIF
         IF (L.LT.NT) GO TO 10
+
 C       -- Newton method for j_{N,L}'
-        IF (N.LE.20) THEN
-           X=0.961587+1.07703*N
+        IF (N.EQ.0) THEN
+           X=3.8317
         ELSE
+C          https://dlmf.nist.gov/10.21#E40
            X=N+0.80861*N**0.33333+0.07249/N**0.33333
         ENDIF
-        IF (N.EQ.0) X=3.8317
+C       iterate
         L=0
-        XGUESS=X
 15      X0=X
         CALL JYNDD(N,X,BJN,DJN,FJN,BYN,DYN,FYN)
         X=X-DJN/FJN
-        IF (X-X0.LT.-1) X=X0-1
-        IF (X-X0.GT.1) X=X0+1
         IF (DABS(X-X0).GT.1.0D-11) GO TO 15
-        IF (L.GE.1)THEN
-           IF (X.LE.RJ1(L)+0.5) THEN
-              X=XGUESS+PI
-              XGUESS=X
-              GO TO 15
-           ENDIF
-        END IF
         L=L+1
         RJ1(L)=X
-C       XXX: should have a better initial guess for large N ~> 100 here
-        X=X+PI+MAX((0.4955d0+0.0915*N-0.000435*N**2)/L, 0d0)
-        IF (L.LT.NT) GO TO 15
+        IF (L.LT.NT) THEN
+C          growth of roots is approximately linear (https://dlmf.nist.gov/10.21#E20)
```

Comment:
Here you’re using not just that the roots of the derivative grow linearly, but that the asymptotic slope is the same as for the roots of the function itself. I think it’s probably easiest to just put the link for the derivative cases.